// bunyip/lib/util.js
var util = require('util');

function extend(target) {
	Array.prototype.forEach.call(arguments, function(arg, i) {
		if (i > 0 && typeof(arg) === 'object') {
			var props = Object.getOwnPropertyNames(arg);

			props.forEach(function(name) {
				var destination = Object.getOwnPropertyDescriptor(arg, name);
				Object.defineProperty(target, name, destination);
			});
		}
	});
	return target;
}

/**
Executes the callback function for each item in array/object. If you return false in the
callback it will break the loop.

@author Moxiecode
@method each
@param {Object} obj Object to iterate.
@param {function} callback Callback function to execute for each item.
 */
function each(obj, callback) {
	var length, key, i;

	if (obj) {
		length = obj.length;

		if (length === undefined) {
			// Loop object items
			for (key in obj) {
				if (obj.hasOwnProperty(key)) {
					if (callback(obj[key], key) === false) {
						return;
					}
				}
			}
		} else {
			// Loop array items
			for (i = 0; i < length; i++) {
				if (callback(obj[i], i) === false) {
					return;
				}
			}
		}
	}
}

/**
Generates an unique ID. This is 99.99% unique since it takes the current time and 5 random numbers.
The only way a user would be able to get the same ID is if the two persons at the same exact milisecond manages
to get 5 the same random numbers between 0-65535 it also uses a counter so each call will be guaranteed to be page unique.
It's more probable for the earth to be hit with an ansteriod. Y

@author Moxiecode
@method guid
@param {String} prefix to prepend (by default 'o' will be prepended).
@method guid
@return {String} Virtually unique id.
 */
var guid = (function() { 
	var counter = 0;
	
	return function(prefix) {
		var guid = new Date().getTime().toString(32), i;

		for (i = 0; i < 5; i++) {
			guid += Math.floor(Math.random() * 65535).toString(32);
		}
		
		return (prefix || '') + guid + (counter++).toString(32);
	}
}());


function uaIdGen(ua) {
	var 
	  map = {
	  	chrome: "Chrome",
	  	firefox: "Firefox",
	  	ie: "IE",
	  	opera: "Opera",
	  	safari: "Safari"
	  }
	, browser = (ua.browser || ua.device || '').toLowerCase()
	, version = parseInt(ua.version, 10)
	;

	if (map[browser]) {
		browser = map[browser];
	}

	return [browser, version].join('|');
}

extend(util, {
	guid: guid,
	uaIdGen: uaIdGen,
	each: each,
	extend: extend,
});

module.exports = util;
